### PR TITLE
Output F from bisection

### PR DESCRIPTION
**Pull request by chengmatt**
Allow F output from bisection (for use in projecting population forward when doing projections).

### CODE REVIEW

File: R/project.R [hunk 1]
```diff
@@ -33,6 +33,7 @@ project <- function(removals, fleet.props, dem_params, prev_naa, recruitment, op
     disc_caa_tmp    = array(NA, dim=c(1, model_params$nages, model_params$nsexes, model_params$nregions, model_params$nfleets))
     caa_tmp         = array(NA, dim=c(1, model_params$nages, model_params$nsexes, model_params$nregions, model_params$nfleets))
     faa_tmp         = array(NA, dim=c(1, model_params$nages, model_params$nsexes, model_params$nregions, model_params$nfleets))
+    F_f_tmp         = array(NA, dim=c(1,  model_params$nregions, model_params$nfleets))
```
Comment (Ovec8hkin) [inline R/project.R:36]: Always use the same dimensions as everything else, even if some of those dimension (e.g. age and sex in this case) seem unnecessary. Different dimensionality makes keeping track of things difficult.

File: R/project.R [hunk 2]
```diff
@@ -68,21 +69,22 @@ project <- function(removals, fleet.props, dem_params, prev_naa, recruitment, op
...
+            naa=prev_naa,
+            fleet.props = fleet.props,
             options=options
         )
 
         land_caa_tmp[,,,r,] <- catch_vars$land_caa_tmp
         disc_caa_tmp[,,,r,] <- catch_vars$disc_caa_tmp
         caa_tmp[,,,r,] <- catch_vars$caa_tmp
         faa_tmp[,,,r,] <- catch_vars$faa_tmp
+        F_f_tmp[,r,] <- catch_vars$F_f
```
Comment (Ovec8hkin) [inline R/project.R:87]: dimensions

File: R/simulate_catch.R [hunk 3]
```diff
@@ -23,6 +23,7 @@ simulate_catch <- function(removals, fleet.props, dem_params, naa, options){
     disc_caa_tmp    <- array(NA, dim=c(1, model_params$nages, model_params$nsexes, 1, model_params$nfleets))
     caa_tmp         <- array(NA, dim=c(1, model_params$nages, model_params$nsexes, 1, model_params$nfleets))
     faa_tmp         <- array(NA, dim=c(1, model_params$nages, model_params$nsexes, 1, model_params$nfleets))
+    F_f_tmp         <- array(NA, dim=c(1, 1, model_params$nfleets))
```
Comment (Ovec8hkin) [inline R/simulate_catch.R:26]: dimensions

File: R/simulate_catch.R [hunk 4]
```diff
@@ -42,7 +43,7 @@ simulate_catch <- function(removals, fleet.props, dem_params, naa, options){
             remove <- removals*fleet.props[f]
 
             # Solve for F that removes catch
-            F_f <- findF_bisection(
+            F_f_tmp[,,f] <- findF_bisection(
```
Comment (Ovec8hkin) [inline R/simulate_catch.R:46]: I would prefer outputting the results of `findF_bisection` to a temporary variable and then assigning to the temporary array at the end. Doesn't really matter at this level, but there is an I/O cost to reading data out of arrays.

**Review by Ovec8hkin — CHANGES_REQUESTED** [review]
Be careful about the dimensionality of temporary arrays for derived quantities. Some quantities naturally have lower dimensionality than full *-at-age arrays, but I think keeping everything at full dimensionality makes it easier to keep track of the dimensions, and allows all of the quantities to play nice with each other in `apply` functions.

**Comment by Ovec8hkin** [comment]
Additionally, why can F not be derived from the FAA matrices, since we know the selectivity curves? @chengmatt

**Comment by chengmatt** [comment]
> Additionally, why can F not be derived from the FAA matrices, since we know the selectivity curves? @chengmatt

Because knowing FAA doesn't necessarily mean you know what F is. You need to know what the apical point of the selectivity curve is to derive F. This then depends on the selectivity of the fishery, which can be different for every stock (e.g., domed; I believe right now, you derive F but summing across 5 age bins or something like that).

**Pull request merged**